cmd/snap-gpio-helper/common.go

```diff
@@ -0,0 +1,235 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2025 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/gadget/device"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/osutil/inotify"
+	"github.com/snapcore/snapd/strutil"
+)
+
+const (
+	aggregatorLockPath         = "/sys/bus/platform/drivers/gpio-aggregator"
+	aggregatorNewDevicePath    = "/sys/bus/platform/drivers/gpio-aggregator/new_device"
+	aggregatorDeleteDevicePath = "/sys/bus/platform/drivers/gpio-aggregator/delete_device"
+	ephermalUdevRulesDir       = "/run/udev/rules.d"
+)
+
+var lockAggregator = func() (unlocker func(), err error) {
+	flock, err := osutil.OpenExistingLockForReading(filepath.Join(dirs.GlobalRootDir, aggregatorLockPath))
+	if err != nil {
+		return nil, err
+	}
+	if err := flock.Lock(); err != nil {
+		return nil, err
+	}
+	return func() {
+		flock.Close()
+	}, nil
+}
+
+var aggregatorCreationTimeout = 120 * time.Second
+var deviceGetGpioChardevChipInfo = device.GetGpioChardevChipInfo
+
+func addAggregatedChip(sourceChip *device.GPIOChardev, commaSeparatedLines string) (chip *device.GPIOChardev, err error) {
```

Comment:
Since this waits, I'd like to pass `context.Context` to it, so that it could be interrupted as well.